src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,558 @@
+// Copyright 2018 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tls
+
+// Delegated credentials for TLS
+// (https://tools.ietf.org/html/draft-ietf-tls-subcerts) is an IETF Internet
+// draft and proposed TLS extension. If the client or server supports this
+// extension, then the server or client may use a "delegated credential" as the
+// signing key in the handshake. A delegated credential is a short lived
+// public/secret key pair delegated to the peer by an entity trusted by the
+// corresponding peer. This allows a middlebox to terminate a TLS connection on
+// behalf of the entity; for example, this can be used to delegate TLS
+// termination to a reverse proxy. Credentials can't be revoked; in order to
+// mitigate risk in case the middlebox is compromised, the credential is only
+// valid for a short time (days, hours, or even minutes).
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"golang.org/x/crypto/cryptobyte"
+)
+
+// DCPeer represents the peer creating or validating a delegated credential.
+type DCPeer uint8
```

Comment:
@armfazh suggested this. So, perhaps check with him ;)

It is exported because in order to create a delegated credential you have to say if you are the server or client. See `NewDelegatedCredential`.

Reply:
An alternative would be simply to pass a `isClient bool` into `NewDelegatedCredential` that indicates whether the DC will be used for client or server authentication. (1) This is consistent with the rest of the codebase; (2) it minimizes the changes to the public API.

Reply:
I added the bool back. If it is not wanted later, I can revert the commit.